File: cmd/hubctl/Privacy.go
// SPDX-License-Identifier: BUSL-1.1
// Copyright (c) 2026 Jérôme Bastian Winkel

package main

import (
	"context"
	"net/url"
	"strconv"
	"time"

	openapitypes "github.com/oapi-codegen/runtime/types"

	"github.com/Jersyfi/hubtask/presentation/openapi"
)

const privacyRequestsPath = "/privacy/requests"

func dsrGroup() group {
	return group{
		name:    "dsr",
		summary: "the rights somebody has exercised, and the deadline on each",
		commands: []command{
			{
				name:    "create",
				usage:   "--kind ACCESS|ERASURE|PORTABILITY|RESTRICTION|OBJECTION|RECTIFICATION [--subject <id>] [--email <address>]",
				summary: "record a case",
				run:     dsrCreate,
			},
			{
				name:    "ls",
				usage:   "[--status <s>] [--kind <k>] [--due-within <days>] [--include-closed]",
				summary: "the open cases, soonest deadline first",
				run:     dsrList,
			},
			{
				name:    "start",
				usage:   "<id> [--mode ANONYMIZE|FULL_DELETE] [--target <id>]",
				summary: "begin the work, which for an access or an erasure is what runs it",
				run:     dsrStart,
			},
			{
				name:    "complete",
				usage:   "<id>",
				summary: "close a case that was answered by hand",
				run:     dsrComplete,
			},
			{
				name:    "reject",
				usage:   "<id> --reason <why>",
				summary: "refuse a case, with the reason - a refusal in time is an answer",
				run:     dsrReject,
			},
		},
	}
}

func dsrCreate(ctx context.Context, cli *CLI, args []string) error {
	flags := commandFlags(cli, "dsr", "create",
		"--kind <kind> [--subject <id>] [--email <address>] [--scope TENANT|INSTALLATION] [--target <id>]")
	kind := flags.String("kind", "", "ACCESS, ERASURE, PORTABILITY, RESTRICTION, OBJECTION or RECTIFICATION")
	subject := flags.String("subject", "", "the account the case is about")
	email := flags.String("email", "", "who asked, for a request with no account behind it")
	scope := flags.String("scope", "", "TENANT, or INSTALLATION for every workspace the person is in")
	target := flags.String("target", "", "the backup target an access or portability export is written to")
	due := flags.String("due", "", "the deadline, when it is not thirty days from receipt")
	notes := flags.String("notes", "", "what was asked, in your own words")
	if err := parseCommand(flags, args); err != nil {
		return err
	}
	if *kind == "" {
		return usagef("dsr create needs --kind")
	}
	if *subject == "" && *email == "" {
		return usagef("dsr create needs --subject or --email: a case is about somebody")
	}

	create := openapi.DataSubjectRequestCreate{Kind: openapi.DataSubjectRequestKind(*kind)}
	if *subject != "" {
		subjectID, err := cli.parseID("--subject", *subject)
		if err != nil {
			return err
		}
		create.SubjectAccountId = &subjectID
	}
	if *email != "" {
		address := openapitypes.Email(*email)
		create.SubjectEmail = &address
	}
	if *scope != "" {
		asked := openapi.DataSubjectRequestScope(*scope)
		create.Scope = &asked
	}
	if *target != "" {
		targetID, err := cli.parseID("--target", *target)
		if err != nil {
			return err
		}
		create.TargetId = &targetID
	}
	if *due != "" {
		deadline, err := cli.parseInstant("--due", *due)
		if err != nil {
			return err
		}
		// parseInstant has already established the form, so this cannot fail; it is parsed again
		// rather than kept as a string because the generated payload carries a time.
		parsed, _ := time.Parse(time.RFC3339, deadline)
		create.DueAt = &parsed
	}
	create.Notes = optional(*notes)

	client, err := cli.client()
	if err != nil {
		return err
	}
	var created openapi.DataSubjectRequest
	if err := client.Post(ctx, privacyRequestsPath, create, &created); err != nil {
		return err
	}
	return cli.Emit(created, caseTable([]openapi.DataSubjectRequest{created}))
}

func dsrList(ctx context.Context, cli *CLI, args []string) error {
	flags := commandFlags(cli, "dsr", "ls",
		"[--status <s>] [--kind <k>] [--due-within <days>] [--include-closed] [--size <n>] [--cursor <c>]")
	status := flags.String("status", "", "RECEIVED, IN_PROGRESS, COMPLETED or REJECTED")
	kind := flags.String("kind", "", "the right that was exercised")
	dueWithin := flags.Int("due-within", 0, "only the cases falling due inside that many days, overdue included")
	includeClosed := flags.Bool("include-closed", false, "the answered ones too")
	size := flags.Int("size", 0, "how many per page")
	cursor := flags.String("cursor", "", "continue the previous page")
	if err := parseCommand(flags, args); err != nil {
		return err
	}

	query := url.Values{}
	for name, value := range map[string]string{"status": *status, "kind": *kind, "cursor": *cursor} {
		if value != "" {
			query.Set(name, value)
		}
	}
	if *dueWithin > 0 {
		query.Set("due_within_days", strconv.Itoa(*dueWithin))
	}
	if *includeClosed {
		query.Set("include_closed", "true")
	}
	if *size > 0 {
		query.Set("size", strconv.Itoa(*size))
	}

	client, err := cli.client()
	if err != nil {
		return err
	}
	var page casePage
	if err := client.Get(ctx, privacyRequestsPath, query, &page); err != nil {
		return err
	}
	if err := cli.Emit(page, caseTable(page.Data)); err != nil {
		return err
	}
	cli.reportMore(page.Page)
	return nil
}

// casePage is the answer to `GET /privacy/requests`: `data` and a `PageInfo`, as every listing in
// this API has, and inline in the contract like the others of this milestone.
type casePage struct {
	Data []openapi.DataSubjectRequest `json:"data"`
	Page openapi.PageInfo             `json:"page"`
}

// dsrStart is the transition that does the work. `RECEIVED → IN_PROGRESS` is what runs the export
// or the erasure, and the job completes the case - so a CLI with `create` and `complete` and
// nothing between them could raise a case and close it having done nothing at all.
func dsrStart(ctx context.Context, cli *CLI, args []string) error {
	const usage = "dsr start <id> [--mode ANONYMIZE|FULL_DELETE] [--target <id>]"
	requestID, rest, err := cli.takeID(args, usage)
	if err != nil {
		return err
	}
	flags := commandFlags(cli, "dsr", "start", "<id> [--mode ANONYMIZE|FULL_DELETE] [--target <id>]")
	mode := flags.String("mode", "", "ANONYMIZE keeps the authorship, FULL_DELETE takes their own contributions too")
	target := flags.String("target", "", "where an access or portability archive is written")
	handledBy := flags.String("handled-by", "", "the account taking the case on")
	if err := parseOnlyFlags(flags, rest, usage); err != nil {
		return err
	}

	update := openapi.DataSubjectRequestUpdate{}
	status := openapi.DataSubjectRequestStatusINPROGRESS
	update.Status = &status
	if *mode != "" {
		erasure := openapi.ErasureMode(*mode)
		update.ErasureMode = &erasure
	}
	if *target != "" {
		targetID, err := cli.parseID("--target", *target)
		if err != nil {
			return err
		}
		update.TargetId = &targetID
	}
	if *handledBy != "" {
		owner, err := cli.parseID("--handled-by", *handledBy)
		if err != nil {
			return err
		}
		update.HandledBy = &owner
	}
	return cli.patchCase(ctx, requestID.String(), update)
}

func dsrComplete(ctx context.Context, cli *CLI, args []string) error {
	const usage = "dsr complete <id>"
	requestID, rest, err := cli.takeID(args, usage)
	if err != nil {
		return err
	}
	flags := commandFlags(cli, "dsr", "complete", "<id>")
	notes := flags.String("notes", "", "what was done, for whoever reads the case later")
	if err := parseOnlyFlags(flags, rest, usage); err != nil {
		return err
	}

	update := openapi.DataSubjectRequestUpdate{Notes: optional(*notes)}
	status := openapi.DataSubjectRequestStatusCOMPLETED
	update.Status = &status
	return cli.patchCase(ctx, requestID.String(), update)
}

func dsrReject(ctx context.Context, cli *CLI, args []string) error {
	const usage = "dsr reject <id> --reason <why>"
	requestID, rest, err := cli.takeID(args, usage)
	if err != nil {
		return err
	}
	flags := commandFlags(cli, "dsr", "reject", "<id> --reason <why>")
	reason := flags.String("reason", "", "why the case is refused; a refusal without one is not an answer")
	if err := parseOnlyFlags(flags, rest, usage); err != nil {
		return err
	}
	if *reason == "" {
		return usagef("dsr reject needs --reason: a refusal within the period is an answer, silence is not")
	}

	update := openapi.DataSubjectRequestUpdate{RejectionReason: reason}
	status := openapi.DataSubjectRequestStatusREJECTED
	update.Status = &status
	return cli.patchCase(ctx, requestID.String(), update)
}

func (cli *CLI) patchCase(
	ctx context.Context, requestID string, update openapi.DataSubjectRequestUpdate,
) error {
	client, err := cli.client()
	if err != nil {
		return err
	}
	var updated openapi.DataSubjectRequest
	if err := client.Patch(ctx, privacyRequestsPath+"/"+requestID, update, &updated); err != nil {
		return err
	}
	return cli.Emit(updated, caseTable([]openapi.DataSubjectRequest{updated}))
}

func caseTable(cases []openapi.DataSubjectRequest) Table {
	rows := make([][]string, 0, len(cases))
	for _, one := range cases {
		rows = append(rows, []string{
			one.Id.String(),
			string(one.Kind),
			string(one.Status),
			subjectOf(one),
			shortTime(&one.DueAt),
			text(one.ResultArchive),
		})
	}
	return Table{
		Columns: []string{"id", "kind", "status", "subject", "due", "archive"},
		Rows:    rows,
	}
}

// subjectOf is who the case is about. The account identifier once there is one, and the address
// otherwise - and a dash after a full deletion, which is exactly right: the case outlives the
// person's record of themselves, and it has to.
func subjectOf(one openapi.DataSubjectRequest) string {
	if one.SubjectAccountId != nil {
		return one.SubjectAccountId.String()
	}
	if one.SubjectEmail != nil && *one.SubjectEmail != "" {
		return *one.SubjectEmail
	}
	return "-"
}

File: apps/webapp/src/lib/data/privacy.ts
// SPDX-License-Identifier: BUSL-1.1
// Copyright (c) 2026 Jérôme Bastian Winkel

/**
 * A right somebody exercised, as a case with a deadline.
 *
 * The pure half, for the reason `jobs.ts` is separate from its store: what a deadline *is* — past,
 * near, or comfortably away — is arithmetic, and it decides how a row is drawn.
 *
 * **A statutory deadline is a due date.** `data-protection.md` §4 gives thirty days from receipt
 * unless the caller names another, and this reads it exactly the way `lib/data/due.ts` reads an
 * entry's: by comparing instants. Inventing a second reading for it would be inventing a second
 * visual language, which is a design decision nobody took.
 *
 * **`INSTALLATION` is not among the scopes here.** It crosses the tenant boundary, needs the
 * `admin:tenants` scope and is the provider's path rather than a tenant's (F4's decision 6). It is
 * absent from the type rather than filtered at the end.
 */

/** The six rights the contract names. */
export type Kind = 'ACCESS' | 'ERASURE' | 'PORTABILITY' | 'RESTRICTION' | 'OBJECTION' | 'RECTIFICATION';

/** Every kind, in the order a picker should offer them. */
export const KINDS: readonly Kind[] = [
  'ACCESS',
  'PORTABILITY',
  'RECTIFICATION',
  'RESTRICTION',
  'OBJECTION',
  'ERASURE',
];

/** The state machine: `RECEIVED → IN_PROGRESS → COMPLETED | REJECTED`. */
export type Status = 'RECEIVED' | 'IN_PROGRESS' | 'COMPLETED' | 'REJECTED';

/** What an erasure does to other people's content. The default is the one that preserves it. */
export type ErasureMode = 'ANONYMIZE' | 'FULL_DELETE';

/** The kinds whose work produces an archive, and which therefore need a target before they start. */
export function producesArchive(kind: Kind): boolean {
  return kind === 'ACCESS' || kind === 'PORTABILITY';
}

/** One case. */
export interface Request {
  readonly id: string;
  readonly kind: Kind;
  readonly status: Status;
  readonly scope?: string;
  readonly subject_account_id?: string | null;
  /** Who asked, for a request that has no account behind it. */
  readonly subject_email?: string | null;
  readonly erasure_mode?: ErasureMode;
  readonly received_at: string;
  /** The statutory deadline. Thirty days from receipt unless the caller named another. */
  readonly due_at: string;
  readonly completed_at?: string | null;
  readonly handled_by?: string | null;
  readonly rejection_reason?: string | null;
  /** Where the export was written at the backup target, once a case has produced one. */
  readonly result_archive?: string | null;
  readonly result_target_id?: string | null;
  readonly notes?: string | null;
}

/**
 * How a deadline stands, in the product's own three states.
 *
 * The same three `DueMark` draws, and for the same reason: a reader who has learned what an overdue
 * task looks like has already learned what an overdue case looks like.
 */
export type Standing = 'overdue' | 'soon' | 'ahead';

/** How near counts as near. Two days, because a statutory answer is not a same-day task. */
export const SOON_MS = 2 * 24 * 60 * 60 * 1000;

/** Where a case's deadline stands at a given moment. A closed case has no standing to report. */
export function standingOf(request: Request, now: number): Standing | undefined {
  if (request.status === 'COMPLETED' || request.status === 'REJECTED') return undefined;
  const due = new Date(request.due_at).getTime();
  if (!Number.isFinite(due)) return undefined;
  if (due <= now) return 'overdue';
  return due - now <= SOON_MS ? 'soon' : 'ahead';
}

/**
 * The cases in the order the list has to be in: closest to its deadline first.
 *
 * The deadline is the column that matters, and a list ordered by anything else would put the case
 * somebody has two days for underneath the one they have three weeks for. Closed cases go last,
 * whatever their deadline was: a deadline that has been answered is not a deadline any more.
 */
export function byDeadline(requests: readonly Request[]): readonly Request[] {
  const closed = (request: Request) =>
    request.status === 'COMPLETED' || request.status === 'REJECTED';
  return [...requests].sort((left, right) => {
    if (closed(left) !== closed(right)) return closed(left) ? 1 : -1;
    return new Date(left.due_at).getTime() - new Date(right.due_at).getTime();
  });
}
